File: Icon.py
class Icon:
    def __init__(self, subtype):
        self.speed = 0.0
        self.glow = 0.0
        self.energy = 0.0
        self.x = 0
        self.y = 0
        self.subtype = subtype  # spinner, slider, or hopper

        # Fields associated with specific subtypes
        self.clockwise = False      # needed for spinner
        self.expand = False         # needed for spinner
        self.vertical = False       # needed for slider
        self.distance = 0           # needed for slider
        self.visible = False        # needed for hopper
        self.xcoord = 0             # needed for hopper
        self.ycoord = 0             # needed for hopper

        # Initialize fields based on subtype
        if self.subtype == 1:
            self.clockwise = True
            self.expand = False
        elif self.subtype == 2:
            self.vertical = True
            self.distance = 0
        elif self.subtype == 3:
            self.visible = True
            self.xcoord = 0
            self.ycoord = 0

    def spin(self):
        print("Spinning...")

    def slide(self):
        print("Sliding...")

    def hop(self):
        print("Hopping...")

    def move(self):
        if self.subtype == 1:
            self.spin()
        elif self.subtype == 2:
            self.slide()
        else:
            self.hop()

    def flair(self):
        if self.subtype == 1:
            self.spin()
        elif self.subtype == 2:
            self.slide()
        else:
            self.hop()

# Example of using the Icon class
my_icon = Icon(1)  # Instantiate with subtype 1 (spinner)
my_icon.move()
my_icon.flair()

File: Icon.c
#include <stdio.h>
#include <stdbool.h>

// Define an enum for better readability of subtypes
typedef enum {
    SPINNER = 1,
    SLIDER,
    HOPPER
} IconSubtype;

typedef struct {
    float speed, glow, energy;
    int x, y;
    int subtype; // spinner, slider, or hopper

    bool clockwise;  // needed for spinner
    bool expand;     // needed for spinner
    bool vertical;   // needed for slider

    int distance;    // needed for slider
    bool visible;    // needed for hopper

    int xcoord, ycoord; // needed for hopper
} Icon;

// Function prototypes
void spin(Icon *icon);
void slide(Icon *icon);
void hop(Icon *icon);
void move(Icon *icon);
void flair(Icon *icon);

// Function to initialize Icon (acts like a constructor)
void init_icon(Icon *icon, int subtype) {
    icon->subtype = subtype;
    // Set initial values based on subtype
    if (subtype == SPINNER) {
        icon->clockwise = true;
        icon->expand = false;
    } else if (subtype == SLIDER) {
        icon->vertical = true;
        icon->distance = 0;
    } else if (subtype == HOPPER) {
        icon->visible = true;
        icon->xcoord = 0;
        icon->ycoord = 0;
    }
}

// Spin function
void spin(Icon *icon) {
    printf("Spinning...\n");
    // Implement spinning logic here
}

// Slide function
void slide(Icon *icon) {
    printf("Sliding...\n");
    // Implement sliding logic here
}

// Hop function
void hop(Icon *icon) {
    printf("Hopping...\n");
    // Implement hopping logic here
}

// Move function
void move(Icon *icon) {
    if (icon->subtype == SPINNER) {
        spin(icon);
    } else if (icon->subtype == SLIDER) {
        slide(icon);
    } else {
        hop(icon);
    }
}

// Flair function
void flair(Icon *icon) {
    if (icon->subtype == SPINNER) {
        spin(icon);
    } else if (icon->subtype == SLIDER) {
        slide(icon);
    } else {
        hop(icon);
    }
}

// Example of using the Icon struct
int main() {
    Icon myIcon;
    init_icon(&myIcon, SPINNER);
    
    move(&myIcon);
    flair(&myIcon);

    return 0;
}
